### PR TITLE
introducing withdrawals (from multisig to target addresses)

### PR DESCRIPTION
**Pull request by monsieurleberre**
these will be used to track the status of transfers from the target chain point of view

### CODE REVIEW

**Comment by monsieurleberre** [comment]
Thanks a lot @fenix2222  good resources and good idea! I was aiming to put a bit more Json specific things in there but the real aim is to do unit tests, I'll change my approach!

File: src/Stratis.FederatedPeg.Features.FederationGateway/Interfaces/IOpReturnDataReader.cs [hunk 1]
```diff
@@ -37,5 +37,12 @@ public interface IOpReturnDataReader
         /// <param name="transaction">The transaction we are examining.</param>
         /// <returns>The address as a string, or null if nothing is found, or if multiple addresses are found.</returns>
         string TryGetTargetAddressFromOpReturn(Transaction transaction);
+
+        /// <summary>
+        /// Tries to find a single OP_RETURN output that can be interpreted as a transaction id.
+        /// </summary>
+        /// <param name="transaction">The transaction we are examining.</param>
+        /// <returns>The transaction id as a string, or null if nothing is found, or if multiple addresses are found.</returns>
```
Comment (zeptin) [inline src/Stratis.FederatedPeg.Features.FederationGateway/Interfaces/IOpReturnDataReader.cs]: Typo, addresses = transaction IDs

Comment (monsieurleberre) [reply src/Stratis.FederatedPeg.Features.FederationGateway/Interfaces/IOpReturnDataReader.cs]: true, thanks

File: src/Stratis.FederatedPeg.Features.FederationGateway/MonitorChain/OpReturnDataReader.cs [hunk 2]
```diff
@@ -142,5 +139,28 @@ public string TryGetTargetAddressFromOpReturn(Transaction transaction)
...
+        public string TryGetTransactionIdFromOpReturn(Transaction transaction)
+        {
+            var transactionId = SelectBytesContentFromOpReturn(transaction)
+                .Select(this.TryConvertValidOpReturnDataToHash)
+                .Where(s => s != null)
+                .Distinct(StringComparer.InvariantCultureIgnoreCase).ToList();
+
+            this.logger.LogDebug("TransactionId found in OP_RETURN(s) of transaction {0}: [{1}]",
+                transaction.GetHash(), string.Join(",", transactionId));
+
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Features.FederationGateway/MonitorChain/OpReturnDataReader.cs]: string.Join happens on transactionId, which suggests there could be more than one.  If so maybe the log should read "TransactionId(s)...".

File: src/Stratis.FederatedPeg.Features.FederationGateway/TargetChain/WithdrawalExtractor.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+        /// <inheritdoc />
+        public IReadOnlyList<IWithdrawal> ExtractWithdrawalsFromBlock(Block block, int blockHeight)
+        {
+            var withdrawals = new List<IWithdrawal>();
+            foreach (var transaction in block.Transactions)
+            {
+                if (transaction.Outputs.Count(o => o.ScriptPubKey != this.withdrawalScript) != 1) continue;
+
+                var targetAddressOutput = transaction.Outputs.Single(o => o.ScriptPubKey != this.withdrawalScript);
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TargetChain/WithdrawalExtractor.cs]: Move this line closer to where the var is referred to.  Above line 52.

File: src/Stratis.FederatedPeg.Features.FederationGateway/Interfaces/IWithdrawal.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+        /// <summary>
+        /// The Id (or hash) of the source transaction that originated the fund
+        /// transfer causing this withdrawal.
+        /// </summary>
+        [JsonConverter(typeof(UInt256JsonConverter))]
+        uint256 Id { get; }
+
+        /// <summary>
+        /// The amount of fund transferred (in target currency).
```
[the remark this answers is not in the document]
Comment (monsieurleberre) [reply src/Stratis.FederatedPeg.Features.FederationGateway/Interfaces/IWithdrawal.cs]: target currency is the currency of the target chain, as opposed to the source one. Conversion is currently 1 to 1.

File: src/Stratis.FederatedPeg.Features.FederationGateway/Interfaces/IOpReturnDataReader.cs [hunk 5]
```diff
@@ -37,5 +37,12 @@ public interface IOpReturnDataReader
         /// <param name="transaction">The transaction we are examining.</param>
         /// <returns>The address as a string, or null if nothing is found, or if multiple addresses are found.</returns>
         string TryGetTargetAddressFromOpReturn(Transaction transaction);
+
+        /// <summary>
+        /// Tries to find a single OP_RETURN output that can be interpreted as a transaction id.
+        /// </summary>
+        /// <param name="transaction">The transaction we are examining.</param>
+        /// <returns>The transaction id as a string, or null if nothing is found, or if multiple ids are found.</returns>
+        string TryGetTransactionIdFromOpReturn(Transaction transaction);
```
[the remark this answers is not in the document]
Comment (monsieurleberre) [reply src/Stratis.FederatedPeg.Features.FederationGateway/Interfaces/IOpReturnDataReader.cs]: OK agreed maybe next PR :)

File: src/Stratis.FederatedPeg.Features.FederationGateway/FederationGatewayFeature.cs [hunk 6]
```diff
@@ -190,15 +195,21 @@ public static class FullNodeBuilderSidechainRuntimeFeatureExtension
...
                     .AddFeature<FederationGatewayFeature>()
                     .DependOn<BlockNotificationFeature>()
                     .FeatureServices(services =>
                     {
+                        services.AddHttpClient(
+                            FederationGatewayFeature.JsonHttpClientName,
+                            client => {
+                                    client.DefaultRequestHeaders.Accept.Clear();
+                                    client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+                            });
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Features.FederationGateway/FederationGatewayFeature.cs]: Now adding the service here, does this make the code in `RestSenderBase.SendAsync()` - lines 32-33 redundant ?

File: src/Stratis.FederatedPeg.Tests/Utils/MaturedBlockDepositBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.VisualStudio.TestPlatform.CommunicationUtilities;
+using NBitcoin;
+
+namespace Stratis.FederatedPeg.Tests.Utils
+{
+    public static class TestingValues
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Tests/Utils/MaturedBlockDepositBuilder.cs]: Class doesn't match the file name.

File: src/Stratis.FederatedPeg.Tests/MaturedBlockDepositTests.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+using Newtonsoft.Json;
+using Stratis.FederatedPeg.Features.FederationGateway.Interfaces;
+using Stratis.FederatedPeg.Features.FederationGateway.Models;
+using Stratis.FederatedPeg.Features.FederationGateway.SourceChain;
+using Stratis.FederatedPeg.Tests.Utils;
+using Xunit;
+
+namespace Stratis.FederatedPeg.Tests
+{
+    public class MaturedBlockDepositModelTests
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Tests/MaturedBlockDepositTests.cs]: Not the same name as the file.

File: src/Stratis.FederatedPeg.Tests/WithdrawalTests.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+using Stratis.FederatedPeg.Features.FederationGateway.Interfaces;
+using Stratis.FederatedPeg.Features.FederationGateway.TargetChain;
+using Stratis.FederatedPeg.Tests.Utils;
+using Xunit;
+
+namespace Stratis.FederatedPeg.Tests
+{
+    public class WithdrawalTests
+    {
+        public static IWithdrawal PrepareWithdrawal()
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Tests/WithdrawalTests.cs]: Private ?

File: src/Stratis.FederatedPeg.Tests/restMaturedBlockSenderTests.cs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+using Microsoft.Extensions.Logging;
+using NSubstitute;
+using NSubstitute.ExceptionExtensions;
+using Stratis.FederatedPeg.Features.FederationGateway;
+using Stratis.FederatedPeg.Features.FederationGateway.SourceChain;
+using Xunit;
+
+namespace Stratis.FederatedPeg.Tests
+{
+    public class RestMaturedBlockSenderTests : IDisposable
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Tests/restMaturedBlockSenderTests.cs]: File name camel case.

File: src/Stratis.FederatedPeg.Tests/restMaturedBlockSenderTests.cs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+            var maturedBlockDeposits = MaturedBlockDepositModelTests.PrepareMaturedBlockDeposits();
+
+            var restSender = new RestMaturedBlockSender(this.loggerFactory, this.federationSettings, this.httpClientFactory);
+
+            await restSender.SendMaturedBlockDepositsAsync(maturedBlockDeposits);
+            this.logger.Received(1).Log<object>(LogLevel.Error, 0, Arg.Any<object>(), Arg.Is<Exception>(e => e != null), Arg.Any<Func<object, Exception, string>>());
+        }
+
+        private void PrepareFailingHttpClient()
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Tests/restMaturedBlockSenderTests.cs]: This and PrepareWorkingHttpClient() could be refactored passing in falling/passing type param. maybe ?

**Pull request merged**